src/instructlab/utils.py

```diff
@@ -564,3 +596,53 @@ def split_hostport(hostport: str) -> tuple[str, int]:
     if not hostname or not port:
         raise ValueError(f"Invalid host-port string: '{hostport}'")
     return hostname, port
+
+
+def convert_messages_to_legacy_dataset(
+    dataset: List[MessageSample],
+) -> List[LegacyMessageSample]:
+    """
+    Converts the new HuggingFace messages dataset format to the legacy format.
+
+    **Note**: The legacy dataset format assumes only a turn of 1. All extra turns will be dropped.
+    This means that we only look at the first 3 messages, and everything afterwards will be ignored.
+    """
+    converted_dataset: List[LegacyMessageSample] = []
+    for dp in dataset:
+        # in new dataset, the roles of a message will be determined.
+        if len(dp["messages"]) < 3:
+            raise ValueError(
+                "Dataset cannot be missing messages from roles. Please make sure that a message from each role is present."
+            )
+
+        converted: LegacyMessageSample = {  # type: ignore
+            m["role"]: m["content"] for m in dp["messages"][:3]
+        }
+        converted_dataset.append(converted)
```

Comment:
nit: There is little use to construct the list only to fail in the middle because one of the datasets doesn't have enough `messages` (line 614). You could probably split into two steps: first validate (and raise if any malformed messages are found); and then calculate / construct the result.

Reply:
What is the concern here about constructing the dataset inline? It would have to be done either way, and the memory will get GC'd once it's no longer being used.

Reply:
The concern is minor because I don't think malformed messages would happen too often.

But in theory, if your last `messages` has `< 3` entries, the function will raise an exception; and all the work to construct `converted_dataset` done so far would be for naught. That's why it could make sense to first validate the input is good; and only then construct `converted_dataset` (you could also then build the result list in one go with a list comprehension instead of calling `.append` on each item - the latter is more expensive.)

The code would then look smth like:

```
if any(len(dp["messages"]) < 3 for dp in dataset):
    raise ...

return [
    {m["role"]: m["content"] for m in dp["messages"][:3]}
    for dp in dataset
]
```

Reply:
While I appreciate your concern around wanting our code to be efficient, I would challenge how expensive this operation is. For example, see the following snippet:

```py
times = []
for k in range(1000):
    t1 = time.time_ns()
    legacy_ds = ensure_legacy_dataset(ds)
    t2 = time.time_ns()
    times.append((t2 - t1) / (1000 ** 3))

average = sum(times) / len(times)
print(f'average time taken: {average:.4f} sec')
print(f"dataset length: {len(ds)}")
print(f"approximate time for a 1,000,000 sample dataset: {average / len(ds) * 1_000_000:.4f} secs")

> average time taken: 0.0121 sec
> dataset length: 24545
> approximate time for a 1,000,000 sample dataset: 0.4947 secs
```

Furthermore, this project revolves around training AI models - which requires a lot of time and computational resource. I would say that it's a much better use of our time to focus on making that more efficient than to try and optimize these one-off operations.

Reply:
That's just a nit, take it or leave it. :)